plutus-benchmark/validation/BenchCek.hs

```diff
@@ -16,11 +17,10 @@ import UntypedPlutusCore as UPLC
      `cabal bench -- plutus-benchmark:validation --benchmark-options crowdfunding`.
 -}
 main :: IO ()
-main = benchWith mkCekBM
+main = evaluate getEvalCtx *> benchWith mkCekBM
```

Comment:
Does this definitely ensure that this is shared across the runs? It seems like it would be simpler to use https://hackage.haskell.org/package/criterion-1.6.0.0/docs/Criterion-Types.html#v:env

Reply:
Ah, or perhaps it wasn't clear: the creation of the evaluation context has non-trivial cost, but _is_ shared by the ledger between script evaluations. So it shouldn't be part of the benchmarked code.

Reply:
Lol, yes, of course it should be `evaluate (force getEvalCtx)`, there's that lazy `Either`.

Reply:
> Did you read the description? The function seems completely irrelevant.

Yes, and I've used it before. I thought the point of it was to clearly compute a value (it lets you use IO so you can do `evaluate`) before the benchmarking begins, and feed it into the individual benchmarks as a function argument. We could even share it across all the benchmarking runs, which would save us some time.

> getEvalCtx is a CAF and doesn't get inlined (it has a NOINLINE pragma and I've verified that it's not getting inlined by looking at the Core anyway). I can't imagine it not being reused, why would GHC ever recompute a non-functional monomorphic top-level value?

This seems like far more complex reasoning than using a function argument and `env`? That guarantees it'll only be computed once.

Reply:
> I thought the point of it was to clearly compute a value (it lets you use IO so you can do evaluate) before the benchmarking begins, and feed it into the individual benchmarks as a function argument.

Reply:
> I thought the point of it was to clearly compute a value (it lets you use IO so you can do evaluate) before the benchmarking begins, and feed it into the individual benchmarks as a function argument.

But it's not what this function is for!

> Motivation. In earlier versions of criterion, all benchmark inputs were always created when a program started running. By deferring the creation of an environment when its associated benchmarks need the its, we avoid two problems that this strategy caused:
>
> - Memory pressure distorted the results of unrelated benchmarks. If one benchmark needed e.g. a gigabyte-sized input, it would force the garbage collector to do extra work when running some other benchmark that had no use for that input. Since the data created by an environment is only available when it is in scope, it should be garbage collected before other benchmarks are run.
> - The time cost of generating all needed inputs could be significant in cases where no inputs (or just a few) were really needed. This occurred often, for instance when just one out of a large suite of benchmarks was run, or when a user would list the collection of benchmarks without running any.

OK, forget about CAF etc. We only care about `getEvalCtx` being forced before benchmarking starts. `evaluate` does exactly that. Once it's forced, it can't be unforced and whatever they do with `env` isn't more reliable than that.

And `getEvalCtx` isn't an input to the benchmarks to even consider `env`. It's a value that the function being benchmarked uses under the hood, you can't feed the value to the function, it's already inside!

If you want to change `evaluateCekLikeInProd` to accept an `EvaluationContext` as an input, I don't mind that, in that case it would make more sense to call `env`, although as I said whatever lazy magic it does, we don't need or care about it, it's completely irrelevant for our use case, so even in that case it would make to force `getEvalCtx` manually.

Reply:
> If you want to change evaluateCekLikeInProd to accept an EvaluationContext as an input

Yes sorry, that's what I was suggesting. `evaluateCekLikeInProd` just calls `getEvalCtx` and then calls `evaluateTerm`. Note that that's actually _less_ like the production version, which really does take it as a parameter and it's cached externally. This way we're relying on some complicated logic that we can force it here and then apparently recompute it inside a function and have it not repeat work. It seems much clearer to just... pass it as an argument!

Reply:
OK then, sounds reasonable.

Feel like doing it yourself? I'm trying to give away as much non-costing-related work as possible.

Reply:
Sure.